jps-plugin/src/org/intellij/erlang/jps/builder/ErlangPrepareDependenciesBuilder.java

```diff
@@ -0,0 +1,270 @@
+    private final LinkedHashMap<String, List<String>> myNodePathsMap = ContainerUtil.newLinkedHashMap();
+
+    public SortedModuleDependencyGraph(@NotNull ErlangDependencyTree dependencyTree) {
+      for (ErlangDependencyTreeNode node : dependencyTree.myNodes) {
+        myNodePathsMap.put(node.myPath, node.myDependencies);
+      }
+    }
+
+    @Override
+    public Collection<String> getNodes() {
+      return myNodePathsMap.keySet();
+    }
+
+    @Override
+    public Iterator<String> getIn(String node) {
+      return myNodePathsMap.get(node).iterator();
+    }
+  }
+
+  private static class BuildOrdersCreator {
+    private final Map<String, SourceRootInfo> mySourceInfoIndex;
+    private final List<ErlangDependencyTreeNode> mySortedNodes;
+
+    public BuildOrdersCreator(@NotNull JpsProject project, @NotNull List<ErlangDependencyTreeNode> sortedNodes) {
+      mySortedNodes = sortedNodes;
+      mySourceInfoIndex = ContainerUtil.newHashMap();
+      for (JpsModule myModule : project.getModules()) {
+        for (JpsTypedModuleSourceRoot<JavaSourceRootProperties> root : myModule.getSourceRoots(JavaSourceRootType.SOURCE)) {
+          mySourceInfoIndex.put(root.getFile().getAbsolutePath(), new SourceRootInfo(myModule.getName(), false));
+        }
+        for (JpsTypedModuleSourceRoot<JavaSourceRootProperties> root : myModule.getSourceRoots(JavaSourceRootType.TEST_SOURCE)) {
+          mySourceInfoIndex.put(root.getFile().getAbsolutePath(), new SourceRootInfo(myModule.getName(), true));
+        }
+      }
+    }
+
+    @NotNull
+    public ErlangModuleBuildOrders createBuildOrderForDirtyFiles(@NotNull Set<String> erlangDirtyFiles) {
+      return createBuildOrderForSortedDirtyModules(getSortedDirtyModules(erlangDirtyFiles));
+    }
+
+    @NotNull
+    private List<String> getSortedDirtyModules(@NotNull final Set<String> allDirtyFiles) {
+      return ContainerUtil.mapNotNull(mySortedNodes, new Function<ErlangDependencyTreeNode, String>() {
+        @Nullable
+        @Override
+        public String fun(ErlangDependencyTreeNode node) {
+          return node.myPath.endsWith(".erl") && allDirtyFiles.contains(node.myPath) ? node.myPath : null;
+        }
+      });
+    }
+
+    private static String getParent(@NotNull String path) {
+      int index = path.lastIndexOf(File.separatorChar);
+      return index > 0 ? path.substring(0, index) : "";
+    }
+
+    @NotNull
+    SourceRootInfo getSourceRootInfoForFile(@NotNull String filePath) {
+      String path = filePath;
+      while (!path.isEmpty() && !mySourceInfoIndex.containsKey(path)) {
+        path = getParent(path);
+      }
+      SourceRootInfo rootInfo = mySourceInfoIndex.get(path);
+      assert rootInfo != null;
+      return rootInfo;
+    }
+
+    @NotNull
+    private ErlangModuleBuildOrders createBuildOrderForSortedDirtyModules(@NotNull List<String> dirtySortedErlangModules) {
+      Map<String, ErlangModuleBuildOrderDescriptor> buildOrders = ContainerUtil.newHashMap();
```

Comment:
You're right. You're not using your XML as sorted data, which is good.

Why not write module build orders to separate XML files, so that we don't parse all build orders every time we build a single module?